### ISSUE TITLE
ModuleBuilder with coarse strategy: primary input not in inputs

### ISSUE THREAD

**Issue by nickclmb** [dart-lang/build#1792]
Outputs of DDC ModuleBuilder have invalid `previousInputsDigest` state.

Module builder with coarse strategy doesn't read its primary input (.dart file) and only reads meta module asset, so when build runner finish build and executes `_setOutputsState`, it removes `primaryInput` asset from `inputs` collection, and calculates MD5 digest for `previousInputsDigest` without primaryInput asset.

But at next build_runner start, AssetGraph deserializer forces adding of `primaryInput` to `inputs` collection:
```dart
for (var output in node.outputs) {
  var inputsNode = graph.get(output) as NodeWithInputs;
  assert(inputsNode != null, 'Asset Graph is missing $output');
  inputsNode.inputs ??= SplayTreeSet<AssetId>();
  inputsNode.inputs.add(node.id);
}
```
As a result, digest of ModuleBuilder nodes is always different on first rebuild.

I assume the fix is to add buildStep.canRead(buildStep.inputId) to module builder with coarse strategy.

**Comment by nickclmb** [comment]
Or build_runner should never remove primaryInput from inputs in `_setOutputsState` function

**Comment by jakemac53** [comment]
This seems like a bug in build_runner - we shouldn't be adding the primary input I think (if you don't read it you shouldn't depend on its bytes).

**Comment by nickclmb** [comment]
Ok, so situation looks like this.
For example for asset   package|path.dart  asset graph will contain:
1 **SourceAssetNode**   id: package|path.dart  
outputs:  package|path.module

2. **GeneratedAssetNode**  id: package|path.module
primaryInput:   package|path.dart
inputs:      package|.meta_module.clean    ~~package|path.dart~~

so the package|path.dart at this moment in inputs of module node after asset graph deserialized, but after first rebuild it will be gone

Doesn't it bring some inconsistency here? I mean  package|path.module is output of some node, but that node is not in inputs?  If thats ok, i will prepare the PR for that issue.

**Comment by jakemac53** [comment]
> Doesn't it bring some inconsistency here? I mean package|path.module is output of some node, but that node is not in inputs? If thats ok, i will prepare the PR for that issue.

The primary input really only determines the outputs, there is no requirement that you read it (although usually you would). As of now even the fine strategy no longer will actually read that file though - it only reads the meta_module asset, and thus the builder shouldn't need to rerun any time the dart file changes, only if the meta_module file changes.

**Comment by jakemac53** [comment]
The actual fix for this might be quite complicated though actually... it might be easiest for now to just treat the primary input as an input always for now.

**Comment by johnpryan** [comment]
for posterity: pinning build_runner to `">=0.10.0 <0.10.1"` seemed to fix the issue for me

**Comment by jakemac53** [comment]
Looks like this is actually an easy fix, putting together a PR now.

**Comment by nickclmb** [comment]
Some measures with and without this fix of this issue:
adding override to one of dependencies, and running serve (with .dart_tool/build cache)

no fix:    [INFO] Build Succeeded after 4m 6s with 7534 outputs (6800 actions)
with fix:  [INFO] Build Succeeded after 45.0s with 10 outputs (7 actions)

**Comment by jakemac53** [comment]
Fix published in build_runner_core version 0.3.1+5
